flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkFunction.java

```diff
@@ -0,0 +1,381 @@
+					long requestOffset = request.getOffset();
+					if (LOG.isDebugEnabled()) {
+						LOG.debug(
+							"Request received, version = " + requestVersion + ", offset = " + requestOffset);
+						LOG.debug(
+							"Expecting version = " + version + ", firstBufferedOffset = " + firstBufferedResultOffset);
+					}
+
+					if (!version.equals(requestVersion) || requestOffset < firstBufferedResultOffset) {
+						// invalid request
+						if (LOG.isDebugEnabled()) {
+							LOG.debug("Invalid request");
+						}
+						sendBufferedResults(0, 0);
+						continue;
+					}
+
+					// valid request, sending out results
+					try {
+						bufferedResultsLock.lock();
+
+						int oldSize = bufferedResults.size();
+						int start = Math.min((int) (requestOffset - firstBufferedResultOffset), oldSize);
+						int end = Math.min(start + maxResultsPerBatch, oldSize);
+
+						if (LOG.isDebugEnabled()) {
+							LOG.debug("Sending back " + (end - start) + " results");
+						}
+						sendBufferedResults(start, end - start);
+
+						if (oldSize >= maxResultsBuffered && bufferedResults.size() < maxResultsBuffered) {
+							bufferNotFullCondition.signal();
+						}
+					} finally {
+						bufferedResultsLock.unlock();
+					}
+				} catch (IOException e) {
+					// IOException occurs, just close current connection
+					// client will come with the same offset if it needs the same batch of results
+					if (LOG.isDebugEnabled()) {
+						LOG.debug("Collect sink server encounters an IOException", e);
+					}
+					closeCurrentConnection();
+				}
+			}
+		}
+
+		private void close() {
+			running = false;
+			closeCurrentConnection();
+			closeServer();
+		}
+
+		private InetSocketAddress getServerSocketAddress() {
+			RuntimeContext context = getRuntimeContext();
+			Preconditions.checkState(
+				context instanceof StreamingRuntimeContext,
+				"CollectSinkFunction can only be used in StreamTask");
+			StreamingRuntimeContext streamingContext = (StreamingRuntimeContext) context;
+			String taskManagerAddress = streamingContext.getTaskManagerRuntimeInfo().getTaskManagerExternalAddress();
+			return new InetSocketAddress(taskManagerAddress, serverSocket.getLocalPort());
+		}
+
+		private InetAddress getBindAddress() {
+			RuntimeContext context = getRuntimeContext();
+			Preconditions.checkState(
+				context instanceof StreamingRuntimeContext,
+				"CollectSinkFunction can only be used in StreamTask");
+			StreamingRuntimeContext streamingContext = (StreamingRuntimeContext) context;
+			String bindHost = streamingContext
+				.getTaskManagerRuntimeInfo()
+				.getConfiguration()
+				.getString(TaskManagerOptions.BIND_HOST);
+
+			if (bindHost != null) {
+				try {
+					return InetAddress.getByName(bindHost);
+				} catch (UnknownHostException e) {
+					return null;
+				}
+			}
+			return null;
+		}
+
+		private void sendBufferedResults(int offset, int length) throws IOException {
+			// skip `offset` results
```

Comment:
I would prefer not to mix the drop & send back logics inside one method, especially the method name is `sendXxxxx`